### ISSUE TITLE
UX: Development process, using IDE, index code

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#964]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/964)
In GitLab by [[Gitlab user @mbodmer]](https://gitlab.com/mbodmer) on Mar 18, 2019, 21:57

## Summary

[//]: # (Summarize the bug encountered concisely)

Using an IDE or Editor with code completion features, the indexer needs to have access to the code and it's dependencies. Also build information is needed. Often a compile_commands.json is available for external tools.

- Why is there no way to access the latest artifact cache from workspaces (dependencies, rootfs, tools)
- How to tell an IDE indexer about the project structure inside the sandbox? compile_commands.json?

What is the envisioned BuildStream workflow for a developer's desktop?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Mar 19, 2019, 12:31

marked this issue as related to #875

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @mbodmer]](https://gitlab.com/mbodmer) on Mar 19, 2019, 14:52

I have stumbled upon this interesting, but discontinued discussion, in buildstream-list: https://mail.gnome.org/archives/buildstream-list/2017-November/msg00046.html

It sounds like things are quite clear and demonstrated by flatpak and flatpak-builder. For me as a user it is however not so clear, how does buildstream deliver support on the points Christian made in his statements?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @mbodmer]](https://gitlab.com/mbodmer) on Mar 19, 2019, 16:31

After a bit more research i have found that #162 has implemented unsafe artifact checkout, which seems to be replaced by ```bst artifact checkout --hardlinks element.bst --directory=sysroot``` now.

This gives me the sysroot which is needed, albeit hard to maintain up to date as a developer. It would be nice if there could be a connection to the workspace, eg an entry in .bstproject.yaml. This could also be parsed by an ide plugin. This connection however should happen automatically in my eyes.

Even if we have the sysroot, there would still be a translation needed between the paths of the buildfiles in the workspace and the real path to the sysroot.

Am I missing something?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 20, 2019, 07:31

Sorry I am late replying this... (loud note to self: there are many people in the project who would have been interested in this topic, but due to gitlab's imperfect notification system, most people don't watch the project for incoming issues).

[[Gitlab user @mbodmer]](https://gitlab.com/mbodmer), there have been discussions in previous hackfests about supporting such features, this should probably be tied to workspaces in some way (so that with an open workspace, one might be able to have an up to date sysroot for an IDE to have access to all of the workspace's dependencies), this might be complicated by the fact that one might likely have multiple workspaces open at the same time.

In short, providing the information to IDEs is very much a desirable feature which has been discussed but currently nobody is working on this as far as I know; [[Gitlab user @coldtom]](https://gitlab.com/coldtom) might have some insight as he is working on a plugin for the GNOME Builder tool.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @mbodmer]](https://gitlab.com/mbodmer) on Mar 20, 2019, 13:38

How could we proceed to help in this situation? If there is a vision maybe people like me could try to improve the situation. But right now I don't really have a clue how an IDE independent integration path could look like.

Following points come to mind (too naive maybe):

- integration through an external tool, library? -> buildstream as a library?
- most IDE's don't have the notion of a build-runtime as gnome-builder has.
- sandbox acts as a remote system (container), some IDE's support remote development (not ideal)
- let buildstream generate a compile_commands.json with sysroot (previous artifact checkout) mapped paths (not sure this would work). many tools and ide's support this format. How can buildstream know details about built parts -> [Bear](https://github.com/rizsotto/Bear)?
- ???

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @coldtom]](https://gitlab.com/coldtom) on Mar 20, 2019, 16:21

[[Gitlab user @adds68]](https://gitlab.com/adds68) is the one working on the plugin for GNOME Builder, not me. I would love to see BuildStream integrated with IDEs though!

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @adds68]](https://gitlab.com/adds68) on Mar 25, 2019, 14:59

marked this issue as related to #972
